File: search_rotated_array.py
# Time complexity : O(log n)
# Space complexity : O(1)
# Leetcode : Solved and submitted
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        
        low = 0
        high = n - 1
        
        # starting the while loop going from low till high
        while low <= high:
          
            # calculate middle element also called the pivot element
            mid = low + (high - low) // 2
            
             # checking if the middle element is the target element, if yes, then return the index
            if target == nums[mid]:
                return mid
              
            # checking if the target element lies within the left sub-array by checking the value of target with the 
            # middle element and the starting element, if yes, then keep searching here, else search in the right subarray
            elif nums[mid] >= nums[low]:
                if target >= nums[low] and target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            
            
            # checking if the target element lies within the right sub-array by checking the value of target with the 
            # middle element and the last element, if yes, then keep searching here, else search in the left subarray
            else:
                if target <= nums[high] and target > nums[mid]:
                    low = mid + 1
                else:
                    high = mid - 1
        
        return -1       

File: src/com/madhurima/SearchRotatedArray.java
/*
    Time Complexity = O(logN)
    Space Complexity = O(1)
    Did this code successfully run on Leetcode : yes
 */

package com.madhurima;

public class SearchRotatedArray {

}

class Solution1 {
    public int search(int[] nums, int target) {

        if(nums == null || nums.length == 0){
            return -1;
        }

        int low = 0;
        int high = nums.length - 1;
        int mid;

        while(low <= high){
            mid = low + (high - low)/2; //to prevent integer overflow
            if(target == nums[mid]){
                return mid;
            }

            if(nums[low] <= nums[mid]){ // left side is sorted
                if(target >= nums[low] && target < nums[mid]){
                    high = mid - 1;
                }else{
                    low = mid + 1;
                }
            }else{ //right side is sorted
                if(target > nums[mid] && target <= nums[high]){
                    low = mid + 1;
                }else{
                    high = mid -1;
                }

            }



        }

        return -1;
    }

    public static void main(String[] args) {
        Solution1 s = new Solution1();
        int[] arr = {4,5,6,7,0,1,2};
        int target = 0;
        System.out.println(s.search(arr, target));
    }
}
